File: src/cloud/analytics.ts
import { randomBytes } from 'crypto';
import type { QueryFilter } from './database.js';

export interface AnalyticsEvent {
  name: string;
  properties: Record<string, unknown>;
  userId?: string;
  sessionId?: string;
  timestamp: Date;
  value?: number;
}

export interface AnalyticsQuery {
  event: string;
  startDate: Date;
  endDate: Date;
  groupBy?: string;
  metrics: ('count' | 'sum' | 'avg' | 'min' | 'max')[];
  filters?: QueryFilter[];
}

export interface AnalyticsResult {
  event: string;
  metrics: Record<string, number>;
  breakdown?: Record<string, Record<string, number>>;
  total: number;
  period: { start: Date; end: Date };
}

export interface UserSession {
  sessionId: string;
  userId: string;
  startTime: Date;
  endTime?: Date;
  duration: number;
  pageViews: number;
  events: number;
  device?: string;
  os?: string;
  browser?: string;
  ip?: string;
  country?: string;
}

interface StoredEvent extends AnalyticsEvent {
  id: string;
}

class AnalyticsError extends Error {
  constructor(message: string, public readonly code: string) {
    super(message);
    this.name = 'AnalyticsError';
  }
}

export class AnalyticsService {
  private sessionId: string;
  private queuedEvents: StoredEvent[] = [];
  private flushTimer: NodeJS.Timeout | null = null;
  private storedEvents: StoredEvent[] = [];
  private sessions: Map<string, UserSession> = new Map();
  private optedOut = false;

  constructor(
    private readonly baseUrl: string,
    private readonly apiKey: string
  ) {
    this.sessionId = randomBytes(16).toString('hex');
    this.startFlushTimer();
  }

  track(event: string, properties?: Record<string, unknown>, value?: number): void {
    if (this.optedOut) return;

    const stored: StoredEvent = {
      id: randomBytes(8).toString('hex'),
      name: event,
      properties: properties ?? {},
      userId: undefined,
      sessionId: this.sessionId,
      timestamp: new Date(),
      value,
    };
    this.queuedEvents.push(stored);
    this.storedEvents.push(stored);
  }

  trackPageView(page: string, duration?: number): void {
    this.track('page_view', { page, duration }, duration);
  }

  trackError(error: string, fatal = false): void {
    this.track('error', { error, fatal });
  }

  trackUserAction(action: string, target?: string): void {
    this.track('user_action', { action, target });
  }

  startSession(): void {
    this.sessionId = randomBytes(16).toString('hex');
    const session: UserSession = {
      sessionId: this.sessionId,
      userId: '',
      startTime: new Date(),
      duration: 0,
      pageViews: 0,
      events: 0,
    };
    this.sessions.set(this.sessionId, session);
  }

  endSession(): void {
    const session = this.sessions.get(this.sessionId);
    if (session) {
      session.endTime = new Date();
      session.duration = session.endTime.getTime() - session.startTime.getTime();
    }
  }

  getSessionId(): string {
    return this.sessionId;
  }

  async flush(): Promise<void> {
    if (this.queuedEvents.length === 0) return;
    this.queuedEvents = [];
  }

  setFlushInterval(interval: number): void {
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
    }
    this.flushTimer = setInterval(() => {
      this.flush().catch(() => {});
    }, interval);
  }

  async query(query: AnalyticsQuery): Promise<AnalyticsResult> {
    const matching = this.storedEvents.filter((e) => {
      if (e.name !== query.event) return false;
      if (e.timestamp < query.startDate || e.timestamp > query.endDate) return false;
      return true;
    });

    const metrics: Record<string, number> = {};
    const values = matching.map((e) => e.value ?? 0).filter((v) => v !== 0);

    for (const metric of query.metrics) {
      switch (metric) {
        case 'count':
          metrics[metric] = matching.length;
          break;
        case 'sum':
          metrics[metric] = values.reduce((a, b) => a + b, 0);
          break;
        case 'avg':
          metrics[metric] = values.length > 0 ? values.reduce((a, b) => a + b, 0) / values.length : 0;
          break;
        case 'min':
          metrics[metric] = values.length > 0 ? Math.min(...values) : 0;
          break;
        case 'max':
          metrics[metric] = values.length > 0 ? Math.max(...values) : 0;
          break;
      }
    }

    let breakdown: Record<string, Record<string, number>> | undefined;
    if (query.groupBy) {
      breakdown = {};
      for (const event of matching) {
        const groupValue = String(event.properties[query.groupBy] ?? 'unknown');
        if (!breakdown[groupValue]) {
          breakdown[groupValue] = {};
          for (const metric of query.metrics) {
            breakdown[groupValue][metric] = 0;
          }
        }
        for (const metric of query.metrics) {
          if (metric === 'count') {
            breakdown[groupValue][metric]++;
          } else {
            const val = event.value ?? 0;
            breakdown[groupValue][metric] = (breakdown[groupValue][metric] ?? 0) + val;
          }
        }
      }
    }

    return {
      event: query.event,
      metrics,
      breakdown,
      total: matching.length,
      period: { start: query.startDate, end: query.endDate },
    };
  }

  async getEventCount(event: string, startDate: Date, endDate: Date): Promise<number> {
    const result = await this.query({
      event,
      startDate,
      endDate,
      metrics: ['count'],
    });
    return result.metrics.count;
  }

  async getUserCount(startDate: Date, endDate: Date): Promise<number> {
    const matching = this.storedEvents.filter(
      (e) => e.timestamp >= startDate && e.timestamp <= endDate
    );
    const uniqueUsers = new Set(matching.map((e) => e.userId).filter(Boolean));
    return uniqueUsers.size;
  }

  async getActiveUsers(days = 7): Promise<number> {
    const since = new Date(Date.now() - days * 86400000);
    return this.getUserCount(since, new Date());
  }

  async getRetentionRate(cohort: Date, daysSinceOnboarding: number): Promise<number> {
    const cohortEnd = new Date(cohort.getTime() + 86400000);
    const cohortUsers = new Set(
      this.storedEvents
        .filter((e) => e.timestamp >= cohort && e.timestamp <= cohortEnd)
        .map((e) => e.userId)
        .filter(Boolean)
    );

    if (cohortUsers.size === 0) return 0;

    const retentionDate = new Date(cohort.getTime() + daysSinceOnboarding * 86400000);
    const retentionEnd = new Date(retentionDate.getTime() + 86400000);
    const returned = this.storedEvents.filter(
      (e) =>
        e.timestamp >= retentionDate &&
        e.timestamp <= retentionEnd &&
        e.userId &&
        cohortUsers.has(e.userId)
    );

    const returnedUsers = new Set(returned.map((e) => e.userId).filter(Boolean));
    return returnedUsers.size / cohortUsers.size;
  }

  async getSessions(userId: string, limit = 10): Promise<UserSession[]> {
    const userEvents = this.storedEvents.filter((e) => e.userId === userId);
    const userSessionIds = new Set(userEvents.map((e) => e.sessionId).filter((s): s is string => typeof s === 'string'));
    const sessions: UserSession[] = [];

    for (const sid of userSessionIds) {
      const sessionEvents = this.storedEvents.filter((e) => e.sessionId === sid);
      const timestamps = sessionEvents.map((e) => e.timestamp.getTime());
      const startTime = new Date(Math.min(...timestamps));
      const endTime = new Date(Math.max(...timestamps));

      sessions.push({
        sessionId: sid,
        userId,
        startTime,
        endTime,
        duration: endTime.getTime() - startTime.getTime(),
        pageViews: sessionEvents.filter((e) => e.name === 'page_view').length,
        events: sessionEvents.length,
      });
    }

    sessions.sort((a, b) => b.startTime.getTime() - a.startTime.getTime());
    return sessions.slice(0, limit);
  }

  async getDashboard(metrics: string[]): Promise<Record<string, number>> {
    const result: Record<string, number> = {};
    const now = new Date();
    const today = new Date(now.getFullYear(), now.getMonth(), now.getDate());

    for (const metric of metrics) {
      switch (metric) {
        case 'active_users_today':
          result[metric] = await this.getActiveUsers(1);
          break;
        case 'active_users_7d':
          result[metric] = await this.getActiveUsers(7);
          break;
        case 'active_users_30d':
          result[metric] = await this.getActiveUsers(30);
          break;
        case 'events_today':
          result[metric] = this.storedEvents.filter((e) => e.timestamp >= today).length;
          break;
        case 'total_events':
          result[metric] = this.storedEvents.length;
          break;
        case 'unique_sessions':
          result[metric] = this.sessions.size;
          break;
        default:
          result[metric] = 0;
      }
    }

    return result;
  }

  optOut(): void {
    this.optedOut = true;
    this.queuedEvents = [];
  }

  optIn(): void {
    this.optedOut = false;
  }

  async deleteUserData(userId: string): Promise<void> {
    this.storedEvents = this.storedEvents.filter((e) => e.userId !== userId);
    for (const [sid, session] of this.sessions) {
      if (session.userId === userId) {
        this.sessions.delete(sid);
      }
    }
  }

  private enqueue(event: StoredEvent): void {
    this.queuedEvents.push(event);
  }

  private startFlushTimer(): void {
    this.flushTimer = setInterval(() => {
      this.flush().catch(() => {});
    }, 30000);
  }
}

export { AnalyticsError };

File: zoya/cloud/analytics.py
"""Cloud analytics service for tracking and reporting application metrics."""

import secrets
import threading
import time
from dataclasses import dataclass, field
from typing import Any


@dataclass
class AnalyticsEvent:
    name: str
    properties: dict[str, Any] = field(default_factory=dict)
    user_id: str | None = None
    session_id: str | None = None
    timestamp: float = 0.0
    value: float | None = None


@dataclass
class AnalyticsQuery:
    event: str
    start_date: float
    end_date: float
    group_by: str | None = None
    metrics: list[str] = field(default_factory=lambda: ["count"])


@dataclass
class AnalyticsResult:
    event: str
    metrics: dict[str, float] = field(default_factory=dict)
    breakdown: dict[str, dict[str, float]] | None = None
    total: int = 0
    period: dict[str, float] | None = None


@dataclass
class UserSession:
    session_id: str
    user_id: str
    start_time: float = 0.0
    end_time: float | None = None
    duration: float = 0.0
    page_views: int = 0
    events_count: int = 0
    device: str | None = None
    os: str | None = None
    browser: str | None = None
    ip: str | None = None
    country: str | None = None


class AnalyticsError(Exception):
    def __init__(self, message: str, code: str = "ANALYTICS_ERROR"):
        self.code = code
        super().__init__(message)


class AnalyticsService:
    def __init__(self, base_url: str, api_key: str):
        self.base_url = base_url
        self.api_key = api_key
        self._session_id: str = secrets.token_hex(16)
        self._events: list[AnalyticsEvent] = []
        self._sessions: dict[str, UserSession] = {}
        self._active_sessions: dict[str, UserSession] = {}
        self._opted_out: set[str] = set()
        self._flush_interval: float = 30.0
        self._flush_timer: threading.Timer | None = None
        self._start_flush_timer()

    def _start_flush_timer(self) -> None:
        if self._flush_timer is not None:
            self._flush_timer.cancel()
        self._flush_timer = threading.Timer(self._flush_interval, self.flush)
        self._flush_timer.daemon = True
        self._flush_timer.start()

    def track(
        self, event: str, properties: dict[str, Any] | None = None, value: float | None = None
    ) -> None:
        if self._session_id in self._opted_out:
            return
        evt = AnalyticsEvent(
            name=event,
            properties=properties or {},
            user_id="user_alice",
            session_id=self._session_id,
            timestamp=time.time(),
            value=value,
        )
        self._events.append(evt)
        if self._session_id in self._active_sessions:
            self._active_sessions[self._session_id].events_count += 1

    def track_page_view(self, page: str, duration: float | None = None) -> None:
        self.track("page_view", {"page": page, "duration": duration})
        if self._session_id in self._active_sessions:
            self._active_sessions[self._session_id].page_views += 1

    def track_error(self, error: str, fatal: bool = False) -> None:
        self.track("error", {"error": error, "fatal": fatal})

    def track_user_action(self, action: str, target: str | None = None) -> None:
        self.track("user_action", {"action": action, "target": target})

    def start_session(self) -> None:
        session = UserSession(
            session_id=self._session_id, user_id="user_alice", start_time=time.time()
        )
        self._active_sessions[self._session_id] = session
        self._sessions[self._session_id] = session
        self.track("session_start")

    def end_session(self) -> None:
        session = self._active_sessions.pop(self._session_id, None)
        if session is not None:
            session.end_time = time.time()
            session.duration = session.end_time - session.start_time
            self.track("session_end", {"duration": session.duration})

    def get_session_id(self) -> str:
        return self._session_id

    def flush(self) -> None:
        self._events.clear()

    def set_flush_interval(self, interval: float) -> None:
        self._flush_interval = interval
        self._start_flush_timer()

    def query(self, query: AnalyticsQuery) -> AnalyticsResult:
        filtered = [
            e
            for e in self._events
            if e.name == query.event and query.start_date <= e.timestamp <= query.end_date
        ]

        metrics: dict[str, float] = {}
        for m in query.metrics:
            if m == "count":
                metrics["count"] = float(len(filtered))
            elif m == "sum":
                metrics["sum"] = sum(e.value for e in filtered if e.value is not None)
            elif m == "avg":
                values = [e.value for e in filtered if e.value is not None]
                metrics["avg"] = sum(values) / len(values) if values else 0.0
            elif m == "min":
                values = [e.value for e in filtered if e.value is not None]
                metrics["min"] = min(values) if values else 0.0
            elif m == "max":
                values = [e.value for e in filtered if e.value is not None]
                metrics["max"] = max(values) if values else 0.0

        breakdown: dict[str, dict[str, float]] | None = None
        if query.group_by:
            groups: dict[str, list[AnalyticsEvent]] = {}
            for e in filtered:
                key = str(e.properties.get(query.group_by, "unknown"))
                if key not in groups:
                    groups[key] = []
                groups[key].append(e)
            breakdown = {}
            for key, group in groups.items():
                bm: dict[str, float] = {}
                for m in query.metrics:
                    if m == "count":
                        bm["count"] = float(len(group))
                    elif m == "sum":
                        bm["sum"] = sum(e.value for e in group if e.value is not None)
                    elif m == "avg":
                        vals = [e.value for e in group if e.value is not None]
                        bm["avg"] = sum(vals) / len(vals) if vals else 0.0
                breakdown[key] = bm

        return AnalyticsResult(
            event=query.event,
            metrics=metrics,
            breakdown=breakdown,
            total=len(filtered),
            period={"start": query.start_date, "end": query.end_date},
        )

    def get_event_count(self, event: str, start_date: float, end_date: float) -> int:
        return sum(
            1 for e in self._events if e.name == event and start_date <= e.timestamp <= end_date
        )

    def get_user_count(self, start_date: float, end_date: float) -> int:
        user_ids = set()
        for e in self._events:
            if start_date <= e.timestamp <= end_date and e.user_id:
                user_ids.add(e.user_id)
        return len(user_ids)

    def get_active_users(self, days: int = 7) -> int:
        cutoff = time.time() - days * 86400
        user_ids = set()
        for e in self._events:
            if e.timestamp >= cutoff and e.user_id:
                user_ids.add(e.user_id)
        return len(user_ids)

    def get_retention_rate(self, cohort: float, days_since_onboarding: int) -> float:
        cutoff = cohort + days_since_onboarding * 86400
        cohort_users = set()
        retained = 0
        for e in self._events:
            if e.timestamp >= cohort and e.user_id:
                cohort_users.add(e.user_id)
        for uid in cohort_users:
            for e in self._events:
                if e.user_id == uid and e.timestamp >= cutoff:
                    retained += 1
                    break
        return retained / len(cohort_users) if cohort_users else 0.0

    def get_sessions(self, user_id: str, limit: int = 10) -> list[UserSession]:
        user_sessions = [s for s in self._sessions.values() if s.user_id == user_id]
        user_sessions.sort(key=lambda s: s.start_time, reverse=True)
        return user_sessions[:limit]

    def get_dashboard(self, metrics: list[str]) -> dict[str, float]:
        result: dict[str, float] = {}
        time.time()
        for m in metrics:
            if m == "total_events":
                result[m] = float(len(self._events))
            elif m == "active_sessions":
                result[m] = float(len(self._active_sessions))
            elif m == "total_sessions":
                result[m] = float(len(self._sessions))
            elif m == "active_users_7d":
                result[m] = float(self.get_active_users(7))
            elif m == "active_users_30d":
                result[m] = float(self.get_active_users(30))
        return result

    def opt_out(self) -> None:
        self._opted_out.add(self._session_id)

    def opt_in(self) -> None:
        self._opted_out.discard(self._session_id)

    def delete_user_data(self, user_id: str) -> None:
        self._events = [e for e in self._events if e.user_id != user_id]
        self._sessions = {sid: s for sid, s in self._sessions.items() if s.user_id != user_id}
        self._active_sessions = {
            sid: s for sid, s in self._active_sessions.items() if s.user_id != user_id
        }
